### PR TITLE
Allow running pre-deployment checks in Kubnernetes

### PR DESCRIPTION
**Pull request by take-five**
This change will allows us to run arbitrary code in Kubernetes on the
new version of the code _before_ deploying the new version. It can be
used for tasks like validating that the new code won't make any harmful
changes is safe to deploy to the environment (think of DB migrations,
applying state to persistent storage etc).

### CODE REVIEW

**Comment by take-five** [comment]
!deploy

**Comment by take-five** [comment]
!deploy

**Comment by sm-deployer** [comment]
@take-five, your input is required [here](https://ci.at.samo.io/job/salemove/job/pipeline-lib/job/PR-107/3/display/redirect) (or [in the old UI](https://ci.at.samo.io/job/salemove/job/pipeline-lib/job/PR-107/3//console)).

**Comment by sm-deployer** [comment]
Deploy failed or was aborted. @take-five, please check the logs [here](https://ci.at.samo.io/job/salemove/job/pipeline-lib/job/PR-107/3/display/redirect) (or [in the old UI](https://ci.at.samo.io/job/salemove/job/pipeline-lib/job/PR-107/3//console)).

**Comment by take-five** [comment]
!deploy

**Comment by sm-deployer** [comment]
@take-five, your input is required [here](https://ci.at.samo.io/job/salemove/job/pipeline-lib/job/PR-107/5/display/redirect) (or [in the old UI](https://ci.at.samo.io/job/salemove/job/pipeline-lib/job/PR-107/5//console)).

**Comment by sm-deployer** [comment]
Deploy failed or was aborted. @take-five, please check the logs [here](https://ci.at.samo.io/job/salemove/job/pipeline-lib/job/PR-107/5/display/redirect) (or [in the old UI](https://ci.at.samo.io/job/salemove/job/pipeline-lib/job/PR-107/5//console)).

**Comment by take-five** [comment]
!deploy

**Comment by sm-deployer** [comment]
@take-five, your input is required [here](https://ci.at.samo.io/job/salemove/job/pipeline-lib/job/PR-107/7/display/redirect) (or [in the old UI](https://ci.at.samo.io/job/salemove/job/pipeline-lib/job/PR-107/7//console)).

**Comment by sm-deployer** [comment]
Deploy failed or was aborted. @take-five, please check the logs [here](https://ci.at.samo.io/job/salemove/job/pipeline-lib/job/PR-107/7/display/redirect) (or [in the old UI](https://ci.at.samo.io/job/salemove/job/pipeline-lib/job/PR-107/7//console)).

**Comment by take-five** [comment]
!deploy

**Comment by sm-deployer** [comment]
@take-five, your input is required [here](https://ci.at.samo.io/job/salemove/job/pipeline-lib/job/PR-107/8/display/redirect) (or [in the old UI](https://ci.at.samo.io/job/salemove/job/pipeline-lib/job/PR-107/8//console)).

**Comment by sm-deployer** [comment]
Deploy failed or was aborted. @take-five, please check the logs [here](https://ci.at.samo.io/job/salemove/job/pipeline-lib/job/PR-107/8/display/redirect) (or [in the old UI](https://ci.at.samo.io/job/salemove/job/pipeline-lib/job/PR-107/8//console)).

**Comment by take-five** [comment]
!deploy

**Comment by sm-deployer** [comment]
@take-five, your input is required [here](https://ci.at.samo.io/job/salemove/job/pipeline-lib/job/PR-107/10/display/redirect) (or [in the old UI](https://ci.at.samo.io/job/salemove/job/pipeline-lib/job/PR-107/10//console)).

**Comment by sm-deployer** [comment]
@take-five, the changes were validated in acceptance. Please click **Proceed** [here](https://ci.at.samo.io/job/salemove/job/pipeline-lib/job/PR-107/10/display/redirect) (or [in the old UI](https://ci.at.samo.io/job/salemove/job/pipeline-lib/job/PR-107/10//console)) to continue the deployment.

**Comment by sm-deployer** [comment]
Deploy failed or was aborted. @take-five, please check the logs [here](https://ci.at.samo.io/job/salemove/job/pipeline-lib/job/PR-107/10/display/redirect) (or [in the old UI](https://ci.at.samo.io/job/salemove/job/pipeline-lib/job/PR-107/10//console)).

File: src/com/salemove/Deployer.groovy [hunk 1]
```diff
@@ -13,7 +13,7 @@ class Deployer implements Serializable {
     \A
     \s*            # Allow optional whitespace before the command
     !deploy
-    (?:            # Don't capture the whitespace before arguments
+    (?:            # Don`t capture the whitespace before arguments
```
Comment (deiwin) [inline src/com/salemove/Deployer.groovy:16]: Why this change?

File: src/com/salemove/Deployer.groovy [hunk 2]
```diff
@@ -398,34 +418,42 @@ class Deployer implements Serializable {
...
+          " ${finalArgs.name}" +
+          " --image='${finalArgs.image}'" +
+          ' --restart=Never' +
+          ' --attach' +
+          ' --rm' +
+          " ${finalArgs.additionalArgs}" +
+          " ${finalArgs.overwriteEntrypoint ? '--command' : ''}" +
+          " -- ${finalArgs.command}",
+        returnStatus: finalArgs.returnStatus,
+        returnStdout: finalArgs.returnStdout
```
Comment (deiwin) [inline src/com/salemove/Deployer.groovy:452]: Could you include these two in the example in the docs as well?

Also, I checked and `returnStatus` does indeed work as one would expect in case of `--restart=Never --attach` as configured here. Note from `kubectl run --help` (emphasis mine):

> --attach=false: If true, wait for the Pod to start running, and then attach to the Pod as if 'kubectl attach ...'
were called.  Default false, unless '-i/--stdin' is set, in which case the default is true. **With '--restart=Never' the exit code of the container process is returned.**

However, does `returnStdout` also work as one would expect. I.e. is stdout clean? Will the command's stdout be returned directly or will there be remnants of `kubectl run` in the output as well?

File: src/com/salemove/Deployer.groovy (hunk 1 above)
Comment (take-five) [reply src/com/salemove/Deployer.groovy:16]: This apostrophe broke syntax highlighting in Github and VS Code (I will push it as a separate commit)

File: src/com/salemove/Deployer.groovy (hunk 2 above)
Comment (take-five) [reply src/com/salemove/Deployer.groovy:452]: Hmm, I didn't think of remnants of kubectl run output. They will be included into result, yes. Do you know from the top of the head how to get rid of it?

Comment (deiwin) [reply src/com/salemove/Deployer.groovy:452]: Maybe they go to stderr and won't be included? If no, then I don't know how to avoid them. Would have to research.

**Comment by take-five** [comment]
!deploy

**Comment by sm-deployer** [comment]
@take-five, the changes were validated in acceptance. Please click **Proceed** [here](https://ci.at.samo.io/job/salemove/job/pipeline-lib/job/PR-107/17/display/redirect) (or [in the old UI](https://ci.at.samo.io/job/salemove/job/pipeline-lib/job/PR-107/17//console)) to continue the deployment.

File: src/com/salemove/Deployer.groovy [hunk 3]
```diff
@@ -379,6 +386,21 @@ class Deployer implements Serializable {
...
+    if (!preDeploymentChecksFor) {
+      script.echo('No pre-deployment checks are defined for this job, skipping.')
+      return
+    }
+
+    script.stage("Running pre-deployment checks in ${env.displayName}") {
+      preDeploymentChecksFor.call(env.subMap(['name', 'domainName']) << [
+        runInKube: { Map args -> runInKube(kubectlCmd, version, args) },
+        version: version,
+        kubectlCmd: kubectlCmd
```
Comment (deiwin) [inline src/com/salemove/Deployer.groovy:399]: Can & should we remove `kubectlCmd` now that we don't really use it? Or do you think the added flexibility might be useful in some other scenarios?

Comment (take-five) [reply src/com/salemove/Deployer.groovy:399]: We can, but I don't think we should. It was totally possible to use `kubectl` during deployment anyway, this helper just makes it less dirty.
